### ISSUE TITLE
Trying port to Lazarus...

### ISSUE THREAD

**Issue by HuguesDug** [Laex/Delphi-OpenCV#48]
Hello,

I tried to port your excellent work to FreePascal / Lazarus. In the most case, the port is easy, except a few problems I could not solve. Would be a great enhancement to FPC/Lazarus and a extension of users to this community.
## Could you integrate these in your next release ?

OCV.Utils, line 406.

  if (isgray) then
    for i := 0 to 255 do
      _rgb[i] := rgb(i, i, i);

_rgb is declared as pCOLORREF, pCOLORREF as ^COLORREF and COLORREF as Cardinal.
Consequently, it does not compile as _rgb is not an array.

Delphi is supporting this, but definitly, does not sound "nice" even to understand what it does.

---

Multiple in units.

Use of {$ELSEIF DEFINED(DelphiOCVVersion_30)} 

$ELSEIF is not declared in FPC.

Should be {$ELSE} {$IFDEF DEFINED(DelphiOCVVersion_30)} .... {$ENDIF}{$ENDIF}

---

Multiple in units.

You use notation in type definitions with <  > 

Example 

Ocv.core.types_c, line 927 
TCV_TREE_NODE_FIELDS<node_type>
or 
Ocv.core.types_c, line 960 
TCV_TREE_NODE_FIELDS<TCvSeq>

The <node_type> does not compile in FPC. Translation is unknown to me. What is the aim at ?

---

Some unit names are not compatible with FPC.

Anyhow, it does not seem to hard to use the rigth names for FPC.

Example in OCV.Lib, we call Winapi.Windows instead if windows. This could be handled with compilar switch.

---

Thank you for taking care... or not. But it would be great

**Comment by zbyna** [comment]
> You use notation in type definitions with < >  ... 

I may be wrong but it seems like generics:
http://sjrd.developpez.com/delphi/tutoriel/generics/?page=introduction#LI-B
fpc support this feature too:
http://www.freepascal.org/docs-html/ref/refch8.html#x102-1240008

**Comment by HuguesDug** [comment]
Hello,

I seached further more. I confirm the <> item should be replaced in FPC compilation with a "generic" keyword front of the type definition. So, "easy done".

Is left the problem of _RGB[i] that I cannot even figure out. I do not even understand how Delphi can accept to compile this (_RGB being a pointer). If target is to go through a memory space, we should see in Inc(_RGB) to move memory block, but I am not sure this is the idea.

**Comment by zbyna** [comment]
Seems like it is connected with:
`{$POINTERMATH ON}`
according to this:
http://rvelthuis.de/articles/articles-pointers.html
it simplifies code:
``` delphi
type 
  PIntegerArray = ^TIntegerArray;
  TIntegerArray = array[0..65535] of Integer;
var
  Buffer: array of Integer;
  PInt: PInteger;
  PArr: PIntegerArray;
  ...
  // Using pointer arithmetic:
  PInt := @Buffer[0];
  for I := 0 to Count - 1 do
  begin
    Writeln(PInt^);
    Inc(PInt);
  end;

  // Using array pointer and indexing:
  PArr := PIntegerArray(@Buffer[0]);
  for I := 0 to Count - 1 do
    Writeln(PArr^[I]);
  ...
end;
 ```
can be written as:
``` delphi
{$POINTERMATH ON}
var
  Buffer: array of Integer;
  PInt: PInteger;
  ...
  // Using new pointer arithmetic:
  PInt := @Buffer[0];
  for I := 0 to Count - 1 do
    Writeln(PInt[I]);
  ...
end;
{$POINTERMATH OFF}
```
So there is no need for the declaration of special TIntegerArray and PIntegerArray types to be able to access the type as an array anymore. Alternatively, instead of PInt[I], the (PInt + I)^ syntax could have been used, with the same result.

**Comment by HuguesDug** [comment]
Super, that works. The compiler directive is switched on with a Compiler Directive reserved for Delphi 2009 top of the unit. Was the cause.

I then found more issues: 
- Some PChar / String convertion problems
- some too many {{ }} that do not close nice
- a problem in assembler part, that return.width instead of TCvSize(Return).width

Now, new "blocking item with pointer arritmetics: 
OCV_Core_Types_C, line 2759. I work on it.

When all fixes will be done, How do I send them back to you ?

**Comment by zbyna** [comment]
I am not developer or owner of this project. I only try to help b/c I use Lazarus and opencv port for it would be great !  :wave: 

In my opinion there is possibility `to fork this project`  here on github and  `patched it`  for being suitable to use with Lazarus. Then you may decide to  `submit a pull request ` of your adjustments  `to the project owner`  back to this project.

**Comment by HuguesDug** [comment]
Hello,

I spent a week-end on it and I do not manage to fix the problem of uses of gererics. 

Ocv.core.types_c, line 927 
TCV_TREE_NODE_FIELDS
or 
Ocv.core.types_c, line 960 
TCV_TREE_NODE_FIELDS

They do not behave the same in FPC and Delphi... so, I am blocked. Any help with this blocking point would help me to move forward.

**Comment by zbyna** [comment]
Hi, 
I had a small accident and I feel pain in my head when using displays so I have to reduce my computer activiries as much as possible for aproximately a week. Then I will try to help with porting.
This may help:
http://www.pascalgamedevelopment.com/showthread.php?6592-Difference-between-FPC-Generics-and-Delphi-Generics
http://wiki.freepascal.org/FPC_New_Features_2.6.0#Generic_records.2C_arrays_and_procedural_types
http://stackoverflow.com/questions/20197179/fpc-tlist-specialization-not-supported

**Comment by zbyna** [comment]
Hi,
i feel better after month. 
I have tried some other ports to lazarus and I realized that in Lazarus is possible to use c api not c++ api (http://answers.opencv.org/question/5421/which-one-is-preferred-to-use-c-or-c/)  B/c there is no clear way how to call c++ api from lazarus see http://rvelthuis.de/articles/articles-cppobjs.html or even no swig support for delphi, I will stick with code::blok IDE and c++ in my small projects. I have no experience to add object pascal support to swig.  Sorry
